src/metabase/app_db/connection_pool_setup.clj

```diff
@@ -27,15 +28,42 @@
   []
   (recent-activity?* @latest-activity recent-window-duration))
 
-(defrecord DbActivityTracker []
+(defn- postgres-connection? [^java.sql.Connection connection]
+  (.isWrapperFor connection org.postgresql.jdbc.PgConnection))
+
+;;; I spun these out into separate functions so we can picc up changes to them in the REPL -- Cam
+
+(defn- on-acquire [_connection]
+  (reset! latest-activity (t/offset-date-time)))
+
+(defn- on-check-in [^java.sql.Connection connection]
+  (reset! latest-activity (t/offset-date-time))
+  ;; for Postgres connections, clean up all resources used in the current session. Otherwise we'll just retain a bunch
+  ;; of crap and it can eat up as much as 300MB per connection. See
+  ;; https://www.postgresql.org/docs/current/sql-discard.html for more info about what this does and
+  ;; https://metaboat.slack.com/archives/C05MPF0TM3L/p1748538414384029?thread_ts=1748507464.051999&cid=C05MPF0TM3L for
+  ;; rationale behind why we're doing it -- Cam
+  (when (postgres-connection? connection)
+    (with-open [stmt (.createStatement connection)]
+      (.execute stmt "DISCARD ALL;"))))
+
+(defn- on-check-out [_connection]
+  (reset! latest-activity (t/offset-date-time)))
+
+(defn- on-destroy [_connection]
+  ;; no-op
+  )
+
+(p/deftype+ MetabaseConnectionCustomizer []
```

Comment:
1. Renamed the type since it's used for more than just activity tracking now
2. Changed from `defrecord` to `deftype` to very slightly reduce the overhead associated with it.
3. Used Potemkin `deftype+` instead of regular `deftype` to prevent redefining the class when reloading the namespace since c3p0 won't pick up the new version of the class after creating the connection pool anyway so no good can come of it
4. Have methods call separate functions so we can iterate on them in the REPL -- otherwise c3p0 wouldn't pick up any changes